### DISCUSSION TITLE
how test Authorization with OAuth2PasswordRequestForm and  custom response body

### DISCUSSION THREAD

**Discussion by XiaobinZhao** [fastapi/fastapi#9004] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### The Question
I  read the the chapter "security" of  fastapi tutorial,and use OAuth2PasswordBearer as the auth function. The code are working successfully with swagger-UI authorize and the try it out. but how to do with testClient? besides，the token api response must return like ‘{"access_token": "xxxx", "token_type": "bearer"}‘， shell we modify it ,just like  ‘{"token": {"access_token": "xxxx", "token_type": "bearer"}}‘ ?

main.py
```Python
from datetime import datetime, timedelta
from typing import Optional

from fastapi import Depends, FastAPI, HTTPException, status
from fastapi.security import OAuth2PasswordBearer, OAuth2PasswordRequestForm
from jose import JWTError, jwt
from passlib.context import CryptContext
from pydantic import BaseModel

# to get a string like this run:
# openssl rand -hex 32
SECRET_KEY = "09d25e094faa6ca2556c818166b7a9563b93f7099f6f0f4caa6cf63b88e8d3e7"
ALGORITHM = "HS256"
ACCESS_TOKEN_EXPIRE_MINUTES = 30


fake_users_db = {
    "johndoe": {
        "username": "johndoe",
        "full_name": "John Doe",
        "email": "johndoe@example.com",
        "hashed_password": "$2b$12$EixZaYVK1fsbw1ZfbX3OXePaWxn96p36WQoeG6Lruj3vjPGga31lW",
        "disabled": False,
    }
}


class Token(BaseModel):
    access_token: str
    token_type: str


class TokenData(BaseModel):
    username: Optional[str] = None


class User(BaseModel):
    username: str
    email: Optional[str] = None
    full_name: Optional[str] = None
    disabled: Optional[bool] = None


class UserInDB(User):
    hashed_password: str


pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

oauth2_scheme = OAuth2PasswordBearer(tokenUrl="auth/token")

app = FastAPI()


def verify_password(plain_password, hashed_password):
    return pwd_context.verify(plain_password, hashed_password)


def get_password_hash(password):
    return pwd_context.hash(password)


def get_user(db, username: str):
    if username in db:
        user_dict = db[username]
        return UserInDB(**user_dict)


def authenticate_user(fake_db, username: str, password: str):
    user = get_user(fake_db, username)
    if not user:
        return False
    if not verify_password(password, user.hashed_password):
        return False
    return user


def create_access_token(data: dict, expires_delta: Optional[timedelta] = None):
    to_encode = data.copy()
    if expires_delta:
        expire = datetime.utcnow() + expires_delta
    else:
        expire = datetime.utcnow() + timedelta(minutes=15)
    to_encode.update({"exp": expire})
    encoded_jwt = jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)
    return encoded_jwt


async def verify_token(token: str = Depends(oauth2_scheme)):
    credentials_exception = HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="Could not validate credentials",
        headers={"WWW-Authenticate": "Bearer"},
    )
    try:
        payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
        username: str = payload.get("sub")
        if username is None:
            raise credentials_exception
        token_data = TokenData(username=username)
    except JWTError:
        raise credentials_exception
    user = get_user(fake_users_db, username=token_data.username)
    if user is None:
        raise credentials_exception
    return user



@app.post("/auth/token", response_model=Token)
async def login_for_access_token(form_data: OAuth2PasswordRequestForm = Depends()):
    user = authenticate_user(fake_users_db, form_data.username, form_data.password)
    if not user:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Incorrect username or password",
            headers={"WWW-Authenticate": "Bearer"},
        )
    access_token_expires = timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
    access_token = create_access_token(
        data={"sub": user.username}, expires_delta=access_token_expires
    )
    return {"access_token": access_token, "token_type": "bearer"}


@app.get("/users/", response_model=User, dependencies=[Depends(verify_token))
async def read_users():
    return fake_users_db["johndoe"]

```
test.py
```python
from fastapi.testclient import TestClient
from myapp.main import app

client = TestClient(app)

def test_login():
    response = client.post("/auth/token", json={"username": "johndoe", "password": "secret", "grant_type": "password"},
                           headers={"content-type": "application/x-www-form-urlencoded"})
    assert response.status_code == 200
    response_data = response.json()
    assert "access_token" in response_data
    assert "token_type" == "bearer"
    return response_data["token_type"], response_data["access_token"]
```
test failed。 the test result is：
```

    def login():
        response = client.post("/auth/token", json={"username": "admin", "password": "password", "grant_type": "password"},
                               headers={"content-type": "application/x-www-form-urlencoded"})
>       assert response.status_code == 200
E       assert 422 == 200
E        +  where 422 = <Response [422]>.status_code
```
1. how to test the login ? 
2. how to custom the login response ?

### Environment

* OS: [Debian 4.9.228]:
* FastAPI 0.63.0:
* python3.7

**Comment by insomnes** [comment, marked as answer]
@XiaobinZhao hello!
> how to test the login ?

You are sending JSON with `"content-type": "application/x-www-form-urlencoded"` that's the problem here, you should send request as form data. Try to change: 
```
response = client.post("/auth/token", data={"username": "johndoe", "password": "secret", "grant_type": "password"},
                           headers={"content-type": "application/x-www-form-urlencoded"})
```
I think this would help

> how to custom the login response ?

You are setting it by yourself with:
`@app.post("/auth/token", response_model=Token)`
and
`return {"access_token": access_token, "token_type": "bearer"}`

But this format is part of Oauth2 specs, so it is better to follow it:
https://tools.ietf.org/html/rfc6749#section-5.1

**Comment by XiaobinZhao** [comment]
@insomnes thanks a lot, your ideas are great.
1. how to test the login ? resolved.
2. how to custom the login response ? i understand the Oauth2 schema and i have login success. but i want wrap my all api response with data,message,error, example '{"data":{ {"access_token": access_token, "token_type": "bearer"}}, "message":"", "error": ""}'. i guess it has to modify swagger-UI ,but i did not find any idear to deal with it still now.

**Comment by pablo-johnson** [comment]
For the ones having this issue and async-asgi-testclient as the client and OAuth2PasswordRequestFormin the route. Use `form` instead of `data`/`json` inside the post method.

Source: https://github.com/vinissimus/async-asgi-testclient/blob/master/async_asgi_testclient/testing.py Line 145
